sbi/neural_nets/mog_flow_snpe_a.py

```diff
@@ -0,0 +1,586 @@
+# This file is part of sbi, a toolkit for simulation-based inference. sbi is licensed
+# under the Affero General Public License v3, see <https://www.gnu.org/licenses/>.
+
+from typing import Union
+from warnings import warn
+
+import torch
+from pyknos.mdn.mdn import MultivariateGaussianMDN
+from pyknos.nflows import flows
+from pyknos.nflows.transforms import CompositeTransform
+from torch import Tensor
+from torch.distributions import MultivariateNormal
+
+import sbi.utils as utils
+from sbi.utils import torchutils
+
+
+class MoGFlow_SNPE_A(flows.Flow):
```

Comment:
Thanks for the PR! 

Just a quick comment regarding the class name: I would suggest replacing `Flow` for `Wrapper` in the name of the class. For a moment, I expected a new flow that works with the analytical division of SNPE-A when reading the PR title and this class name in the description. Replacing `Flow` for `Wrapper` would change emphasis to it being a wrapper to make the interfaces compatible, avoiding this potential confusion.

Reply:
I agree, but are you sure you want to move the whole class to the snpe_a.py file? If we do this, it would be the only case where the implementation of a density estimator would be outside the neural_nets folder. As a result, `get_nn_models.py` in `utils` would need to import from `snpe_a.py` which itself loads `utils`.

Reply:
I agree, that's indeed not great.

How about this alternative (maybe I'm missing something though):
1) in `mdn_snpe_a.py`, instead of the line
```
neural_net = MoGFlow_SNPE_A(transform, distribution, embedding_net)
```
we write
```
neural_net = flows.flow(transform, distribution, embedding_net)
```
2) Instead of `transform, distribution, embedding_net`, `MoGFlow_SNPE_A` takes the `neural_net` and sets it as an attribute `self.net`. Then, instead of inheritance, we always use `self.net.something` instead of `super().something()`
3) Lastly, in `sbi/inference/snpe/snpe_a.py` lines 140-160 (when the neural net is initialized), we write a wrapper such as
```
if ...
    self._build_base_neural_net = partial(self._build_neural_net, num_components=1) # Same for the other cases
elif ...
else...
def wrap_mdn_for_snpe_a(theta):
    return MoGFlow_SNPE_A(self._build_base_neural_net(theta))

self._build_neural_net = wrap_mdn_for_snpe_a
```

I hope this was kinda clear, please let me know I wrote confusing or wrong things ;)

Reply:
@jan-matthis that's a good suggestion. We will rename it accoringly.

Reply:
(if you want this PR to be merged rather quickly and move on (which is totally understandable), we can also merge it without this big refactor and I can try out some things once it's merged).

Reply:
thank you for the offer. how about we fix some easy things by today evening, such that you can merge it tomorrow. there are still a few of your first iteration comments that we want to realize

Reply:
Based on your suggestion, we did ma big rework of the MoGWrapper_SNPE_A (also renamed to MDNWrapper_SNPE_A). It is now only constructed when building a posterior, and takes the trained flow and a proposal as input. This way, we could delete the custom factory function. Thanks @theogruner!

Reply:
Ah, yes, that's great!

Let me know when you're done with all fixes, then I'll merge it tomorrow :)